Control/Parallel/Strategies.hs

```diff
@@ -527,48 +530,67 @@ evalBufferWHNF n0 xs0 = return (ret xs0 (start n0 xs0))
            start !n  (y:ys) = y `pseq` start (n-1) ys
 
 -- | 'evalBuffer' is a rolling buffer strategy combinator for (lazy) lists.
+-- Pattern matching on the result will evaluate the first @n+1@ elements using
+-- the given strategy. Pattern matching on each additional list cons will
+-- evaluate an additional element.
 --
--- 'evalBuffer' is not as compositional as the type suggests. In fact,
--- it evaluates list elements at least to weak head normal form,
--- disregarding a strategy argument 'r0'.
---
--- > evalBuffer n r0 == evalBuffer n rseq
+-- @evalBuffer n strat = buffering n (rseq `dot` strat)@
 --
+-- 'evalBuffer' is not as compositional as the name and type suggest.
+-- In fact, it evaluates list elements at least to weak head normal
+-- form, disregarding a strategy argument of 'r0'. For a compositional
+-- version, see 'buffering'.
 evalBuffer :: Int -> Strategy a -> Strategy [a]
-evalBuffer n strat =  evalBufferWHNF n . map (withStrategy strat)
-
--- Like evalBufferWHNF but sparks the list elements when pushing them
--- into the buffer.
--- Not to be exported; used in parBuffer and for optimisation.
-parBufferWHNF :: Int -> Strategy [a]
-parBufferWHNF n0 xs0 = return (ret xs0 (start n0 xs0))
+evalBuffer n0 strat xs0 = return (ret xs0 (start n0 xs0))
   where -- ret :: [a] -> [a] -> [a]
-           ret (x:xs) (y:ys) = y `par` (x : ret xs ys)
+           ret (x:xs) (y:ys) = withStrategy strat y `pseq` (x : ret xs ys)
            ret xs     _      = xs
 
         -- start :: Int -> [a] -> [a]
            start 0   ys     = ys
            start !_n []     = []
-           start !n  (y:ys) = y `par` start (n-1) ys
-
-
--- | Like 'evalBuffer' but evaluates the list elements in parallel when
--- pushing them into the buffer.
-parBuffer :: Int -> Strategy a -> Strategy [a]
-parBuffer n strat = parBufferWHNF n . map (withStrategy strat)
--- Alternative definition via evalBuffer (may compromise firing of RULES):
--- parBuffer n strat = evalBuffer n (rparWith strat)
+           start !n  (y:ys) = withStrategy strat y `pseq` start (n-1) ys
+-- Alternative definition via evalBufferWHNF (won't be deforested):
+-- evalBuffer n strat =  evalBufferWHNF n . map (withStrategy strat)
 
--- Deforest the intermediate list in parBuffer/evalBuffer when it is
--- unnecessary:
+-- Avoid unnecessary withStrategy calls in/evalBuffer:
 
 {-# NOINLINE [1] evalBuffer #-}
-{-# NOINLINE [1] parBuffer #-}
 {-# RULES
 "evalBuffer/rseq"  forall n . evalBuffer  n rseq = evalBufferWHNF n
-"parBuffer/rseq"   forall n . parBuffer   n rseq = parBufferWHNF  n
```

Comment:
I think it was pretty important to optimise this case, I'm worried this might be a regression.

Reply:
Ideally, users wouldn't write `parBuffer n rseq`, because that's just the same as `parBuffer n r0`. I think the right way to catch this (and other such) is probably to write rules for the underlying operations:

```haskell
runEval (r0 x) ==> x
r0 (runEval (rseq x)) ==> r0 x
r0 (runEval (rpar x)) ==> r0 x
rseq (runEval (rpar x)) ==> rseq x
rpar (runEval (rseq x)) ==> rpar x
rpar (runEval (rpar x)) ==> rpar x
rseq (runEval (rseq x)) ==> rseq x
```

Making the higher-order combinators `INLINABLE` should then expose things sufficiently to these low-level rules.

Reply:
Note if you're following by email: I edited the above comment substantially.

Reply:
I added a comment explaining what `tieConses` does. I don't think it's actually complicated. I wrote it with `foldr` so we can consider making the whole thing `INLINABLE` and get list fusion on one side. Perhaps it would be easier to see what's going on if you write it with explicit recursion?

Reply:
Could you explain why `parBuffer n rseq` should be the same as `parBuffer n r0`? That seems counter-intuitive to me.

Reply:
It's really because `parBuffer` isn't compositional. A better way is to use the `runEval (rpar x) = x` rule so the strategy is simplified before being passed to `buffering`. I can change the PR. I'm wondering if you would be okay with redefining `evalBuffering` to `buffering`; it's a breaking change, but makes things much more uniform across the module.

Reply:
@simonmar, um, sorry, what I just said was a bit confused. The *current* behavior of `parBuffer` actually equates `parBuffer n r0` to `parBuffer n rseq`. My implementation in terms of `buffering` and `rparWith` does not, once `rparWith` is fixed.